secrets/file_test.go

```diff
@@ -1,217 +1,169 @@
+		origin := t.TempDir() + "/origin"
+
+		path := t.TempDir() + "/foo"
+		require.NoError(t, os.Symlink(origin, path))
+
+		assert.Error(t, sp.Add(path))
+	})
+
+	t.Run("refreshes symlink to file", func(t *testing.T) {
+		origin := t.TempDir() + "/origin"
+		writeFile(t, origin, "created")
+
+		path := t.TempDir() + "/foo"
+		require.NoError(t, os.Symlink(origin, path))
+
+		require.NoError(t, sp.Add(path))
+
+		_, exists := sp.GetSecret(origin)
+		assert.False(t, exists)
+
+		checkSecret(t, path, "created")
+
+		writeFile(t, origin, "updated")
+		checkSecret(t, path, "updated")
+
+		removeFile(t, origin)
+		_, exists = sp.GetSecret(path)
+		assert.True(t, exists)
+
+		writeFile(t, origin, "re-created")
+		checkSecret(t, path, "re-created")
+	})
+
+	t.Run("ignores subdirectories", func(t *testing.T) {
+		dir := t.TempDir()
+
+		require.NoError(t, os.Mkdir(dir+"/subdir", 0700))
+
+		writeFile(t, dir+"/foo", "created")
+		writeFile(t, dir+"/subdir/bar", "ignored")
+
+		require.NoError(t, sp.Add(dir))
+
+		checkSecret(t, dir+"/foo", "created")
+
+		_, exists := sp.GetSecret(dir + "/subdir")
+		assert.False(t, exists)
+
+		_, exists = sp.GetSecret(dir + "/subdir/")
+		assert.False(t, exists)
+
+		_, exists = sp.GetSecret(dir + "/subdir/bar")
+		assert.False(t, exists)
+	})
+
+	t.Run("trims newline", func(t *testing.T) {
+		path := t.TempDir() + "/foo"
+
+		writeFile(t, path, "created\n")
+
+		require.NoError(t, sp.Add(path))
+
+		checkSecret(t, path, "created")
+	})
 }
 
-func Test_SecretPaths_Close(t *testing.T) {
-	temproot, err := os.MkdirTemp(os.TempDir(), "skipper-secrets-close")
-	if err != nil {
-		t.Errorf("Failed to create temp dir: %v", err)
-		return
-	}
-	defer func() {
-		t.Logf("remove %s", temproot)
-		os.RemoveAll(temproot)
-	}()
-	watchit := temproot + "/watch"
-	os.MkdirAll(watchit, 0777)
-	dat := []byte("data")
-	afile := watchit + "/afile"
-
-	sp := NewSecretPaths(60 * time.Millisecond)
-	err = os.WriteFile(afile, []byte(""), 0644)
-	if err != nil {
-		t.Errorf("Failed to create file: %v", err)
-	}
-	err = sp.Add(afile)
-	if err != nil {
-		t.Errorf("Failed to Add file to watch list: %v", err)
-	}
-	err = os.WriteFile(afile, dat, 0644)
-	if err != nil {
-		t.Errorf("Failed to write to file: %v", err)
-	}
-	time.Sleep(100 * time.Millisecond) // wait for refresher
+func TestSecretPathsDoesNotRefreshAfterClose(t *testing.T) {
+	sp := NewSecretPaths(refresh)
 
-	got, ok := sp.GetSecret(afile)
-	if !ok {
-		t.Errorf("Should have secret: %v", ok)
-	}
-	if string(got) != string(dat) {
-		t.Errorf("Secret content, got: %s, want: %s", string(got), string(dat))
+	checkSecret := func(t *testing.T, path string, expected string) {
+		got, ok := sp.GetSecret(path)
+		if assert.True(t, ok) {
+			assert.Equal(t, []byte(expected), got)
+		}
 	}
 
+	path := t.TempDir() + "/foo"
+
+	writeFile(t, path, "created")
+
+	require.NoError(t, sp.Add(path))
+
+	checkSecret(t, path, "created")
+
 	sp.Close()
```

Comment:
I guess after Close you need to synchronize or time.Sleep to not be open to the closing race: ` Close -> close(sp.quit) -> for { select { case <-sp.quit: return }}` 
and ` writeFile`

Reply:
yes, looks like it occasionally writes into the file too fast; added a delay after close